src/core/ext/transport/chttp2/transport/frame_data.cc

```diff
@@ -104,23 +104,22 @@ grpc_error* grpc_deframe_unprocessed_incoming_frames(
     uint8_t* end = nullptr;
     uint8_t* cur = nullptr;
 
-    grpc_slice slice = grpc_slice_buffer_take_first(slices);
-
-    beg = GRPC_SLICE_START_PTR(slice);
-    end = GRPC_SLICE_END_PTR(slice);
+    grpc_slice* slice = grpc_slice_buffer_peek_first(slices);
+    beg = GRPC_SLICE_START_PTR(*slice);
+    end = GRPC_SLICE_END_PTR(*slice);
     cur = beg;
     uint32_t message_flags;
     char* msg;
 
     if (cur == end) {
-      grpc_slice_unref_internal(slice);
+      grpc_slice_buffer_consume_first(slices);
```

Comment:
It's almost like this needs some kind of RAII... but I'm ok with that being a followup PR

Reply:
I was thinking of the same, but a bit more than RAII. What we need IMHO is a position iterator. Something that basically iterates over the bytes of a slice buffer, and provides a way to tell slice buffer to consume up to a given position. But, that really needs a humongous change, separate from this patch. :-)